### ISSUE TITLE
Trouble replicating GPT-2 on KV260

### ISSUE THREAD

**Issue by kjw-54321** [itsnsreehari/KV260-Accelerated-Platform#1]
Hi, thanks for sharing this project!

I'm trying to replicate your setup, but I'm running into some issues—especially with the hardware and GPT-2 parts.

**Hardware**

I’m using the KV260 board and tried loading the DPU bitstream:

ubuntu@kria:~/hardware$ sudo xmutil loadapp dpu_b4096.xclbin
dpu_b4096.xclbin: load Error: -1

I used the kv260.xclbin file from the GitHub repo and renamed it to dpu_b4096.xclbin. Not sure if that's correct, but it fails to load.

**GPT-2 Quantization**

I'm also confused about how to quantize GPT-2 for deployment. I don’t see an .xmodel file in the repo, so I assume I need to generate it myself. I’ve tried using PyTorch quantization and ONNX quantion in vitis ai docker to quantize gpt2 from transformers, but I’m a bit lost on the exact steps to go from the original model to a deployable xmodel for the KV260.

Would really appreciate any guidance or a quick overview of the steps you followed.

Thanks!

**Comment by itsnsreehari** [comment]
Hi, thanks so much for looking at the project!

About the hardware problem:
Yes, simply renaming the .xclbin won't cut it. xmutil requires a certain metadata format that is provided when the bitstream is packaged in the correct way as a Vitis platform application. In my environment, I made the kv260.xclbin in Vivado + Vitis flow specifically for KV260 board, and exported it correctly using package_xo and v++ linking. You can't simply rename a generic.xclbin and hope it'll work. apologies, I should have made that clearer in the README.

On GPT-2 Quantization:
You are absolutely correct. It's not as simple as it sounds! I began from HuggingFace's gpt2 model and performed a lightweight distillation + quantization in PyTorch initially. But Vitis AI doesn't directly support GPT-like models by default, so I manually adjusted the model to a form friendly to the DPU (particularly in the areas surrounding MatMul ops). Subsequently, I exported to ONNX, and then performed vai_q_onnx for quantization and then vai_c_xir for generating the.xmodel. For vai_c_xir compilation you would need your hardware file (.xclbin / bit.bin) generated using the Vitis Hardware flow and the fingerprint of your arch.json should match the fingerprint of the model. This is absolutely necessary. 

It's a little hacky and most certainly not plug-and-play. If you're attempting to reproduce this, I'd recommend beginning with something like a smaller distilled one such as distilgpt2 and reducing the attention layers initially.

If you find yourself stuck on a certain step, don't hesitate to send me a message. I'll be more than happy to assist!

Thanks again :)

**Comment by kjw-54321** [comment]
**hardware problem:**
I used the kv260.xclbin from your GitHub and followed the steps in the README. Just to double-check, are the xclbin and bit.bin files in the github what you used for your setup, or is it a generic file? If it’s not the one you used, would you mind sharing how you generated the hardware files for the KV260?  That part's still a bit unclear to me.

**GPT-2 Quantization:**
Thanks a lot for explaining your process! I’ll definitely try starting with distilgpt2 and simplifying the model to better fit the DPU, like you suggested. If it’s not too much trouble, could you share a bit more detail on the specific steps or commands you used for vai_q_onnx and vai_c_xir to generate the .xmodel? That would really help me get unstuck.

Sorry for all the questions—I'm pretty new to Vitis AI, but your project looked so cool I really wanted to try it out. Thanks again for being so generous with your time and help!

**Comment by itsnsreehari** [comment]
Hey, no worries at all, really glad you're trying this out and happy to help!

# Hardware Setup:

Yeah, the kv260.xclbin is the exact one I used for my final setup. To generate your own hardware files (so that there's a fingerprint match) for the KV260, here’s the general flow I followed: 

1. Start with this official Xilinx tutorial for creating the platform:
https://github.com/Xilinx/Vitis-Tutorials/blob/2022.2/Vitis_Platform_Creation/Design_Tutorials/01-Edge-KV260/README.md
Follow it till you’ve created the Vivado project and exported the .xpfm platform. ( Doing these steps to generate the hardware files would be really helpful especially if you're just staring out on Vitis and Vivado). 


2. Once your platform is ready, use the steps from this guide (up to the “Build Hardware Overlay” section):
https://www.hackster.io/LogicTronix/zcu102-vitis-dpu-trd-vitis-ai-3-0-c51609
That will help you generate a proper .xclbin file that works with xmutil. ( Steps are just the same even though the guide is for ZCU102. When you run the Vitis Hardware Development Flow after importing the xpfm from the last step, it would generate the desired .xclbin and arch.json. This arch.json would contain your unique fingerprint which needs to be used later on during the compilation process. 



For the DPU architecture, I used the B4096 config, it’s optimized for larger matrix ops like what GPT-2 requires, though it might still need model simplification. You’ll configure this during the DPU IP generation stage in Vivado.


# GPT-2 Quantization & .xmodel Generation:

Here’s the process I used:

1. Start with a smaller model like distilgpt2 from HuggingFace.


# Export to ONNX:
Use the HuggingFace transformers and optimum libraries to export:

`from transformers import AutoTokenizer, AutoModelForCausalLM
from optimum.exporters.onnx import main_export`


`main_export("distilgpt2", output="onnx_model", task="text-generation")`


# Quantize the ONNX model:
Use vai_q_onnx inside the Vitis AI Docker:

`vai_q_onnx quantize \
  --model onnx_model/model.onnx \
  --calib_data calib.txt \
  --output_dir quantized_model \
  --input_nodes input_ids \
  --output_nodes logits \
  --input_shapes [[1,16]]  # Adjust depending on your max sequence length`


# Compile to XModel using vai_c_xir:

`vai_c_xir \
  --xmodel quantized_model/deploy_model.onnx \
  --arch /opt/vitis_ai/compiler/arch/DPUCZDX8G/KV260/arch.json \
  --output_dir compiled_model`

(You would get a compilation error and the compiled model wouldn't be generated if there's a fingerprint mismatch. Hence, it's recommended you create the hardware files from scratch :)


Some minor model tweaking might be needed depending on attention layers or unsupported ops. I had to strip down and refactor some parts manually to make it DPU-compatible.


Let me know how it goes! You’re doing great, Vitis AI has a learning curve, but once you get through the first build, things start making more sense. Happy to help further if you run into roadblocks.

**Comment by kjw-54321** [comment]
Thank you so much for your detailed explanation. It’s been incredibly helpful in getting me this far with the KV260 setup. I wanted to share my current progress and ask for a bit more guidance on the next step.

So far, I’ve:
-    Exported distilgpt to ONNX and quantized using Vitis AI docker
-    Followed the official Xilinx tutorial to generate the KV260 hardware platform (.xsa) and exported the .xpfm.
-    Set up a PetaLinux 2024.2 project using that .xsa.

Now I am working on following this tutorial. It says if i created a custom platform, i need to create the common image file myself too.
Currently I am trying to build a Linux image to match the platform, but kept running into issues: 
        petalinux-image-minimal fails due to an automatic dependency on libvcu-omxil, 

![Image](https://github.com/user-attachments/assets/20504db2-541c-437f-b639-effdf4178736)

Also, even if I manage to create the image, I’m stumped on Section B. Building the Hardware Overlay. It’s not clear where exactly should I be running the make commands, and what those are. Would i just need to do the same thing on the logs provided below on the site?

If I could get a bit more clarity on how you followed the tutorial, especially around building the .xclbin, it  would really help me.

🔗 https://www.hackster.io/LogicTronix/zcu102-vitis-dpu-trd-vitis-ai-3-0-c51609

Thanks so much

**Comment by itsnsreehari** [comment]
Hi! Thanks a lot for the detailed update. You're making great progress and it's awesome to hear the KV260 setup is coming along well.

Regarding your current issues, here’s a structured response to guide you further:

# Pre-Built PetaLinux and Ubuntu Images

You do not need to build the Linux image from scratch unless you have a very specific requirement. The official, pre-built PetaLinux and Ubuntu images for the KV260 and other Kria SOMs are readily available here:

Kria SOMs Starter Kits - Official Pre-Built Images - https://xilinx-wiki.atlassian.net/wiki/spaces/A/pages/1641152513/Kria+SOMs+Starter+Kits

These include:

common_image-kv260

Ubuntu-based smartcam starter images

PetaLinux-based minimal boot images


This should help you avoid dependency errors like the one you're seeing with libvcu-omxil.



# Hardware Overlay and make Command

Yes, to build the hardware overlay (.xclbin), you’ll need to run the make command inside this directory:

DPUCZDX8G_VAI_v3.0/prj/Vitis

This will initiate the build process and generate the xclbin file required for your DPU design.

If you're wondering where to get the DPUCZDX8G_VAI_v3.0 directory, it is typically part of the official DPU IP package from AMD/Xilinx. You can search and download it from the AMD developer site:  https://www.amd.com/en/developer.html
Just look for the Vitis AI 3.0 release or DPUCZDX8G IP package (you may need to create an AMD account to access it).

Once you have the DPUCZDX8G in place, you can change the DPU architecture in the dpu_config.vh as mentioned in the blog. After you've figured out the directory, you would just need to run those four commands mentioned in the logs. This would automatically generate the binary file (.xclbin) after you run the make command. It would take a couple of hours depending upon the DPU architecture (took me 4.5 hrs for b4096 on intel 13th gen i7). You would be able to see all the files except the SD card image because we aren't doing it for ZCU102. 

# A Few Tips

Ensure your .xpfm and the compiled platform (.xsa) are consistently used across Vitis and PetaLinux builds.

When working inside the Vitis AI Docker, make sure you're mapping your platform path correctly so that it recognizes your custom platform.

The final xclbin will be required for inference deployment using VART.


Best of luck, you're almost there!

**Comment by kjw-54321** [comment]
Hi again, thanks for your continued help!

I’ve made some progress but still have a few questions related to the Deployment Instructions

**dpu.xclbin:**

I successfully generated the xclbin and .xmodel files with your help. However, when trying to load the xclbin file, I still get the following error:

    dpu.xclbin: load Error: -1

I followed the steps you gave me, but the same error happens. and it doesnt show any detailed error messages, so i am quite lost


**pl.dtbo**

I generated the pl.dtbo file using the xsa I used for making xclbin. However, when I tried to add the pl.dtbo using the dtbo-manager, I encountered this error:

sudo: dtbo-manager: command not found

Could you advise on the proper way to add the pl.dtbo to the system, or suggest an alternative to dtbo-manager?

**shell.json**
While I have successfully created the xclbin and .xmodel files, I am unsure what the shell.json file is and how it’s used. Is this file something I need to generate manually, or is it created during the DPU configuration process? Any guidance on this would be greatly appreciated.

Thanks again for your time and help!

**Comment by itsnsreehari** [comment]
Hi! Thanks for the update. glad to see you’ve made solid progress! I'll address your issues one by one: 

# dpu.xclbin Load Error: load Error: -1

This error usually indicates a mismatch between the running hardware design and the xclbin you're trying to load. Please check the following:

- Make sure that the bitstream used during boot (BOOT.BIN) and the platform (.xsa) used to generate your xclbin are the same.

- (Important) Confirm that the correct dpu.xclbin is placed in `/lib/firmware/xilinx/<your_platform>/` and that your shell.json is pointing to the right path. If the .xclbin is not placed at `/lib/firmware/xilinx/<your_platform>/` the file will fail to load resulting in the load Error -1. This is most probably the solution to your load issue. 

- After copying your dpu.xclbin, don’t forget to run:

  `sudo xmutil unloadapp`
   `sudo xmutil loadapp <your_app>`

- If the system still throws load Error: -1, try checking dmesg:

  `dmesg | grep xocl`

This might reveal more detail about why the xclbin failed to load (e.g., clock mismatch, resource binding issues).


# pl.dtbo and dtbo-manager not found

The dtbo-manager utility is typically available only in the official Kria PetaLinux and Ubuntu images. If you’re using a minimal or custom PetaLinux image, it might not be included.

Here are alternatives:

Instead of using dtbo-manager, you can manually overlay the device tree blob:

  `mkdir /sys/kernel/config/device-tree/overlays/pl`
  `echo /lib/firmware/pl.dtbo > /sys/kernel/config/device-tree/overlays/pl/path`

Note: Make sure the configfs is mounted:

  `mount -t configfs configfs /sys/kernel/config`

If you want to persist the overlay across boots, consider appending the .dtbo directly into your boot image or using a script to apply it on startup.


# shell.json: 

shell.json is a metadata configuration file used by the Vitis AI runtime (VART) to understand your deployment environment. It tells the runtime which DPU kernel(s) are available, where the xclbin is, and what device to target.

You can either:

- Use an example shell.json from the official DPU TRD (e.g., from DPUCZDX8G_VAI_v3.0/overlay/kv260/shell.json) and modify it with your paths and kernel name.

- Or generate it yourself manually. A minimal example:

`{
  "target": "DPUCZDX8G",
  "xclbin": "/lib/firmware/xilinx/kv260_custom_platform/dpu.xclbin",
  "kernel": "DPUCVDX8H",  // Use the correct kernel name used in your design
  "device": "0"
}`

In simpler terms: Create a text file in Linux and paste the above contents. 

Make sure this file is placed where your application or VART runtime can access it, or specify its path when launching the app.


You're really close to the final deployment. Keep it up!
